File: 알고리즘_도감/sgoldenbird/week4.js
// 1. 최초 등장 값 찾기
function solution1(arr, target) {
  for (let i of arr) {
    if(i === target) return arr[i]; 
  }
  return -1; 
}

// 2. 정렬된 배열에서의 값 위치 찾기
function solution2(nums, target) {
  let left = 0;
  let right = nums.length-1;

  while (left <= right){
    const mid = Math.floor((left+right)/2);

    if (nums[mid] === target){
      return mid;
    }else if (nums[mid] < target){
      left = mid+1; // 탐색 범위를 오른쪽 절반으로 좁힙니다.
    }else {
      right = mid-1; // 탐색 범위를 왼쪽 절반으로 좁힙니다.
    }
  }

  return left;  // 반복문이 끝날 때까지 target을 찾지 못했다면, left 포인터가 가리키는 위치가 target이 삽입되어야 할 올바른 인덱스가 됩니다.
}

// 3. 주어진 노드로부터 연결된 노드 개수 세기
function solution3(adj, startNode) {
    return adj[startNode].length;
}


// 4. 가장 가까운 연결된 노드 찾기
function solution4(adj, startNode, targetNode) {
  if (startNode === targetNode) return 0;

  // 방문 여부를 기록할 배열을 만듭니다. (노드 개수만큼 false로 초기화)
  const visited = new Array(adj.length).fill(false);
    
  // 큐를 생성하고 [현재 노드 번호, 현재까지의 경로 길이] 형태로 시작 노드를 넣습니다.
  const queue = [[startNode, 0]];
  visited[startNode] = true; // 시작 노드 방문 표시

  // 큐가 빌 때까지 반복합니다.
  while (queue.length > 0) {
    // 가장 앞에 있는 노드 정보를 꺼냅니다.
    const [currentNode, currentDistance] = queue.shift();

    // 현재 꺼낸 노드가 목표 노드라면, 이때의 거리가 최단 거리이므로 즉시 반환합니다.
    if (currentNode === targetNode) {
        return currentDistance;
    }

    // 현재 노드와 직접 연결된 이웃 노드들을 살펴봅니다.
    const neighbors = adj[currentNode];
    for (let i = 0; i < neighbors.length; i++) {
      const nextNode = neighbors[i];

      // 아직 방문하지 않은 이웃 노드만 큐에 넣고 방문 표시를 합니다.
      if (!visited[nextNode]) {
          visited[nextNode] = true;
          // 다음 노드로 한 칸 이동하므로 거리를 +1 해줍니다.
          queue.push([nextNode, currentDistance + 1]);
      }
    }
  }

  return -1;
}


// 5. 그래프 경로 존재 여부 확인
function solution5(adj, startNode, targetNode) {
 
  const visited = new Set();

  function dfs(currentNode) {

      if (currentNode === targetNode) {
          return true;
      }

      // 현재 노드를 이미 방문한 적이 있다면 더 이상 들어가지 않고 종료합니다.
      if (visited.has(currentNode)) {
          return false;
      }

      visited.add(currentNode);

      const neighbors = adj[currentNode];
      for (let i = 0; i < neighbors.length; i++) {
          const nextNode = neighbors[i];
          
          // 이웃 노드로 깊숙이 들어갔을 때(재귀 호출) true가 반환된다면
          // 결국 목표 노드를 찾았다는 뜻이므로, 아래에 남은 다른 이웃들은 보지도 않고 true를 위로 쭉 올립니다.
          if (dfs(nextNode)) {
              return true;
          }
      }

      // 모든 이웃을 타고 끝까지 들어가 봤는데도 목표를 못 찾았다면 false를 반환합니다.
      return false;
  }

  // 준비된 재귀 함수에 시작 노드를 넣고 실행한 결과를 그대로 반환합니다.
  return dfs(startNode);
}

File: 알고리즘_도감/gnsals0904/week4.java
import java.util.LinkedList;
import java.util.Queue;

class week4 {
    public static void main(String[] args) {
        int[][] adj = new int[][] {
            {1, 2},
            {0, 3},
            {0},
            {1}
        };

        System.out.println(solution1(new int[] {3, 1, 4, 1, 5, 9, 2, 6}, 1));
        System.out.println(solution2(new int[] {1, 3, 5, 6}, 5));
        System.out.println(solution3(adj, 0));
        System.out.println(solution4(adj, 0, 3));
        System.out.println(solution5(adj, 0, 3));
    }

    /**
     * 1. 최초 등장 값 찾기
     *   - 배열을 처음부터 순서대로 확인
     *   - target을 처음 발견한 인덱스를 반환
     *   - 끝까지 찾지 못하면 -1을 반환
     * @param arr
     * @param target
     * @return
     */
    static int solution1(int[] arr, int target) {
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == target) {
                return i;
            }
        }

        return -1;
    }

    /**
     * 2. 정렬된 배열에서의 값 위치 찾기
     *   - 이진 탐색으로 탐색 범위를 절반씩 줄임
     *   - target이 있으면 해당 인덱스를 반환
     *   - target이 없으면 left가 삽입될 위치가 됨
     * @param nums
     * @param target
     * @return
     */
    static int solution2(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;
        int minIndex = nums.length;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (nums[mid] >= target) {
                minIndex = Math.min(minIndex, mid);
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return minIndex;
    }

    /**
     * 3. 주어진 노드로부터 연결된 노드 개수 세기
     *   - 인접 리스트에서 startNode 위치의 배열 길이를 확인
     *   - 직접 연결된 노드 개수를 반환
     * @param adj
     * @param startNode
     * @return
     */
    static int solution3(int[][] adj, int startNode) {
        return adj[startNode].length;
    }

    /**
     * 4. 가장 가까운 연결된 노드 찾기
     *   - BFS로 가까운 노드부터 탐색
     *   - targetNode에 도착하면 현재 거리를 반환
     * @param adj
     * @param startNode
     * @param targetNode
     * @return
     */
    static int solution4(int[][] adj, int startNode, int targetNode) {
        boolean[] visited = new boolean[adj.length];
        Queue<Integer> queue = new LinkedList<>();

        queue.add(startNode);
        visited[startNode] = true;
        int distance = 0;

        while (!queue.isEmpty()) {
            int size = queue.size();

            for (int i = 0; i < size; i++) {
                int currentNode = queue.poll();

                if (currentNode == targetNode) {
                    return distance;
                }

                for (int nextNode : adj[currentNode]) {
                    if (!visited[nextNode]) {
                        visited[nextNode] = true;
                        queue.add(nextNode);
                    }
                }
            }

            distance++;
        }

        return -1;
    }

    /**
     * 5. 그래프 경로 존재 여부 확인
     *   - DFS로 한 방향을 최대한 깊게 탐색
     *   - targetNode를 찾으면 true
     *   - 모든 경로를 확인해도 찾지 못하면 false
     * @param adj
     * @param startNode
     * @param targetNode
     * @return
     */
    static boolean solution5(int[][] adj, int startNode, int targetNode) {
        boolean[] visited = new boolean[adj.length];
        return dfs(adj, startNode, targetNode, visited);
    }

    static boolean dfs(int[][] adj, int currentNode, int targetNode, boolean[] visited) {
        if (currentNode == targetNode) {
            return true;
        }

        visited[currentNode] = true;

        for (int nextNode : adj[currentNode]) {
            if (!visited[nextNode] && dfs(adj, nextNode, targetNode, visited)) {
                return true;
            }
        }

        return false;
    }
}
